File: services/payment-service/app/models/processed_event.py
"""Markers for events this service has already acted on.

Kafka delivers at least once, so every consumer has to recognise a redelivery. A row here
is written in the same transaction as the business change it guards: both commit or
neither does. See "Consuming events" in docs/events/README.md.
"""

import uuid
from datetime import datetime

from sqlalchemy import DateTime, String, Uuid, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, utcnow


class ProcessedEvent(Base):
    __tablename__ = "processed_events"

    # (event_id, consumer) rather than event_id alone: two handlers in one service may each
    # need to process the same event once.
    event_id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True)
    consumer: Mapped[str] = mapped_column(String(64), primary_key=True)
    processed_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=utcnow, server_default=func.now()
    )

File: services/order-service/src/OrderService.Infrastructure/Persistence/ProcessedEvent.cs
namespace OrderService.Infrastructure.Persistence;

/// <summary>
/// One row of <c>processed_events</c>: this consumer has acted on this event.
/// </summary>
/// <remarks>
/// A persistence detail rather than a domain concept, so it lives here and not in the
/// domain. The key is <c>(EventId, Consumer)</c> so two handlers in this service could each
/// process the same event once.
/// </remarks>
public sealed class ProcessedEvent
{
    private ProcessedEvent()
    {
        // Required by EF Core.
        Consumer = string.Empty;
    }

    public ProcessedEvent(Guid eventId, string consumer, DateTimeOffset processedAt)
    {
        EventId = eventId;
        Consumer = consumer;
        ProcessedAt = processedAt;
    }

    public Guid EventId { get; private set; }

    public string Consumer { get; private set; }

    public DateTimeOffset ProcessedAt { get; private set; }
}
